sdk/cpp/core/src/netconf_tcp_client.cpp

```diff
@@ -0,0 +1,444 @@
+/*  ----------------------------------------------------------------
+ Copyright 2016 Cisco Systems
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------*/
+
+#include <cstdio>
+#include <map>
+#include <string.h>
+#include <sstream>
+#include <string>
+#include <unistd.h>
+#include <vector>
+
+#include <curl/curl.h>
+#include <libxml/parser.h>
+#include <libxml/tree.h>
+
+#include "capabilities_parser.hpp"
+#include "errors.hpp"
+#include "ietf_parser.hpp"
+#include "logger.hpp"
+#include "netconf_tcp_client.hpp"
+
+
+namespace ydk
+{
+static const long TIMEOUT = 6000L;
+static const int EIGHT_K = 8196;
+
+static const char EOM_10[] = "]]>]]>";
+static const char EOM_11[] = "\n##\n";
+static const char LF_HASH[] = "\n#";
+static const char LF[] = "\n";
+
+
+static const std::map<const std::string, const std::string> PATTERN_MAP{{"&quot;", "\""},
+                                                                       {"&amp;", "&"},
+                                                                       {"&apos;", "'"},
+                                                                       {"&lt;", "<"},
+                                                                       {"&gt;", ">"},
+                                                                       {"\n##\n", ""}};
+
+static const std::string NETCONF_11("urn:ietf:params:netconf:base:1.1");
+
+static const char *HELLO_11 = "\n<hello xmlns=\"urn:ietf:params:xml:ns:netconf:base:1.0\">\n"
+                              "<capabilities>\n"
+                              "<capability>urn:ietf:params:netconf:base:1.1</capability>\n"
+                              "</capabilities>\n"
+                              "</hello>\n"
+                              "]]>]]>";
+
+
+static int wait_on_socket(curl_socket_t sockfd, int for_recv, long timeout_ms);
+static bool ends_with_framing(const char* buf, size_t nread, int version);
+static void xml_to_string(xmlDocPtr doc, xmlNodePtr root, std::string &out);
+static xmlDocPtr get_xml_doc(const std::string &payload);
+static std::string trim_chunk(std::string str);
+static std::string trim_reply(std::string str);
+static void replace(std::string& subject, const std::string& search, const std::string& replace);
+
+
+NetconfTCPClient::NetconfTCPClient(std::string username, std::string password,
```

Comment:
Can these variables be changed from `std::string` to `const std::string &` to avoid copying?